snuba/replacer.py

```diff
@@ -381,3 +410,85 @@ def flush_batch(self, batch: Sequence[Replacement]) -> None:
             logger.info(
                 "Optimized %s partitions on %s" % (num_dropped, clickhouse_read.host)
             )
+
+    def _message_already_processed(self, metadata: ReplacementMessageMetadata) -> bool:
+        """
+        Figure out whether or not the message was already processed.
+
+        Check if there exists a recorded offset that took too long to execute.
+        If there is, we can conclude whether or not to process the incoming
+        message by comparing the incoming message to the recorded offset.
+        """
+        key = self._build_topic_group_index_key(metadata)
+        self._reset_offset_check(key)
+
+        # float("-inf") implies this is the first message this worker is processing since startup
+        if self.__prolonged_offset == float("-inf"):
+            processed_offset = redis_client.get(key)
+            try:
+                self.__prolonged_offset = (
+                    -1 if processed_offset is None else int(processed_offset)
+                )
+            except ValueError as e:
+                redis_client.delete(key)
+                logger.warning(
+                    "Unexpected value found for an offset in Redis", exc_info=e,
+                )
+
+        return metadata.offset <= self.__prolonged_offset
+
+    def _reset_offset_check(self, key: str) -> None:
+        """
+        We may need to clear the offset the replacer is comparing incoming messages
+        against.
+
+        Eg. The offset is manually reset in Kafka to start processing messages
+        again from an older offset. The replacer will ignore this and just skip
+        messages till it's back to the offset stored in Redis.
+
+        There exists a config which tells us which consumer groups require their
+        replacer(s) reset. Ideally this config is populated with consumer groups
+        temporarily, then cleared once relevant consumers restart.
+        """
+        # expected format is "[consumer_group1, consumer_group2, ..]"
```

Comment:
This is a bit of a nit but the comment here has spaces after the commas. I assume that's not accurate?

Reply:
You're right, force of habit 🤦‍♂️